master/pkg/searcher/asha.go

```diff
@@ -170,7 +172,30 @@ func (s *asyncHalvingSearch) trialClosed(
 ) ([]Operation, error) {
 	s.TrialsCompleted++
 	s.ClosedTrials[requestID] = true
-	return nil, nil
+	var ops []Operation
+	if s.sendShutdownOperation() {
+		allTrials := len(s.Rungs[0].Metrics)
+		log.Warnf("Shutdown initiated; failures: %d, trials requested: %d", s.InvalidTrials, allTrials)
+		shutdown := Shutdown{Failure: s.InvalidTrials >= allTrials}
+		ops = append(ops, shutdown)
+	}
+
+	return ops, nil
+}
+
+func (s *asyncHalvingSearch) sendShutdownOperation() bool {
+	shutdown := true
+	allValidTrials := len(s.Rungs[0].Metrics)
+	allCreatedTrials := len(s.Rungs[0].Metrics) + s.InvalidTrials
+	if s.PendingTrials > 0 || allCreatedTrials > s.TrialsCompleted || allValidTrials < s.MaxTrials() {
```

Comment:
s.PendingTrials > 0  and allValidTrials < s.MaxTrials() conditions are similar to random and grid search. 

The check allCreatedTrials > s.TrialsCompleted has to be made to ensure that all the Close Operations are processed (even for Invalid Trials) otherwise there might be some left over Close operations that aren't processed and searcher will shutdown prematurely. This is different from others because in ASHA NewClose is not created right after NewCreate as you'll see in random and grid. 

OutStandingTrials for each rung counts any trial created in the bottom rung or promoted to other rungs. The summation of OutstandingTrials in all rungs should always be <= s.PendingTrials. So, I think we can remove this.  Any thoughts?